In [20]:
import pandas as pd
from operator import attrgetter
from datetime import timedelta

In [21]:
customers = pd.read_csv("olist_customers_dataset.csv")
#таблица с уникальными идентификаторами пользователей

In [22]:
orders = pd.read_csv("olist_orders_dataset.csv", parse_dates=[3, 4, 5, 6, 7])
#таблица заказов

In [23]:
order_items = pd.read_csv("olist_order_items_dataset.csv")
#товарные позиции, входящие в заказы

In [24]:
customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [25]:
customers.shape

(99441, 5)

In [26]:
customers.customer_unique_id.nunique()
#уникальных пользователей

96096

In [27]:
customers.customer_id.nunique()
#уникльных заказов

99441

In [28]:
customers.isna().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [29]:
orders.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0                    2017-10-18  
1                    2018-08-13  
2                    2018-09-04  
3                    2017-12-15  
4                    2018-02-26

In [30]:
orders.order_purchase_timestamp.min()

Timestamp('2016-09-04 21:15:19')

In [31]:
orders.order_purchase_timestamp.max()

Timestamp('2018-10-17 17:30:18')

In [32]:
orders.shape

(99441, 8)

In [33]:
orders.isna().sum()
#Не везде есть даты, что логично: не все заказы успешно завершены

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [34]:
orders.customer_id.nunique()
#уникльных пользователей (равно числу заказов в customers; их столько же, сколько order_id) но на самом деле заказов

99441

In [35]:
orders.order_id.nunique()
#всего заказов

99441

In [36]:
orders.groupby("order_status", as_index = False).agg({"customer_id": "count"})
#взглянем на распределение заказов по статусам

order_status  customer_id
0     approved            2
1     canceled          625
2      created            5
3    delivered        96478
4     invoiced          314
5   processing          301
6      shipped         1107
7  unavailable          609

In [37]:
order_items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [38]:
order_items.order_id.nunique()
#где ещё 775 чеков?

98666

In [39]:
order_items.shape

(112650, 7)

In [40]:
order_items.isna().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

In [41]:
order_items.product_id.nunique()
#наименований товаров

32951

In [42]:
order_items.order_item_id.nunique()
#максимальное колво позиций в одном чеке

21

In [43]:
order_items.groupby("order_id", as_index = False).agg({"order_item_id": "count"}).sort_values("order_item_id", ascending=False).head()
#смотрим, сколько позиций товара в каждом чеке: от 1 до 21

order_id  order_item_id
50137  8272b63d03f5f79c56e9e4120aec44ef             21
10459  1b15974a0141d54e36626dca3fdc731a             20
65715  ab14fdcfbe524636d65ee38360e22ce8             20
60941  9ef13efd6949e4573a18964dd1bbe7f5             15
25583  428a2f660dc84138d969ccd69a0ab6d5             15

In [44]:
customers_orders = customers.merge(orders)

# 1. Сколько у нас пользователей, которые совершили покупку только один раз? (7 баллов) 

Как понять вообще, что пользователь совершил покупку? Давайте взглянем на статусы и подумаем, как их соотнести с фактом покупки.
    created —  создан - очевидно, человек просто набрал карзину товаров и нажал "заказать", купил ли он это? Если посмотреть на все статусы, то мы видем целый жизненный цикл заказа и делаем вывод, что нажав заказат далеко не уедешь.
    approved —  подтверждён - значит, сайт увидел заказ, пока рано говорить о совершённой сделке
    invoiced —  выставлен счёт - вооот,тут уже две стороны сработались и идут дальше, но если заказ не оплатить, то покупа е свершится. Идём дальше.
    processing —  в процессе сборки заказа - вот теперь мы имеем такую картину: кто-то нажал заказать, мы подтвердили, что видим, попросили оплатить, т.к. сл. статус "shipped", предполагаем, что оплата прошла успешно и мы деньги получили, теперь можем смело говорить, что покупка совершна. Дальше отгрузка и доставка мб вообще не в нашей юрисдиции, м.б. мы отпрвляем почтой или пользуемся услугами иной логистической фирмы.
    shipped —  отгружен со склада
    delivered —  доставлен пользователю
    unavailable —  недоступен
    canceled —  отменён
    
Итого, будем считать, что все заказы со статусами: processing, shipped, delivered - оплачены и не отменены, соответственно, пользователь совершил покупку.

In [45]:
customer_1_buy = customers_orders.query("order_status == ['processing', 'shipped', 'delivered']") \
                .groupby("customer_unique_id", as_index=False) \
                .agg({"customer_zip_code_prefix": "count"}) \
                .query("customer_zip_code_prefix == 1").shape[0]

In [46]:
customer_1_buy

91814

# 2. Сколько заказов в месяц в среднем не доставляется по разным причинам (вывести детализацию по причинам)? (5 баллов)

In [47]:
#Недоставленные согласно статусу заказа
orders.query("order_status != 'delivered'").shape[0]

2963

In [48]:
#Недоставленные, т.к. отсутствует дата доставки
orders.order_delivered_customer_date.isna().sum()

2965

Значения при разных подходах получились практически равные, но при этом с разными заказами, попробуем ещё покрутить

In [49]:
#Недоставленные, т.к. отсутствует дата доставки - выводим их
orders[orders.order_delivered_customer_date.isna()].head()

order_id                       customer_id  \
6    136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
44   ee64d42b8cf066f35eac1cf57de1aa85  caded193e8e47b8362864762a83db3c5   
103  0760a852e4e9d89eb77bf631eaaf1c84  d2a79636084590b7465af8ab374a8cf5   
128  15bed8e2fec7fdbadb186b57c46c92f2  f3f0e613e0bdb9c7cee75504f0f90679   
154  6942b8da583c2f9957e990d028607019  52006a9383bf149a4fb24226b173106f   

    order_status order_purchase_timestamp   order_approved_at  \
6       invoiced      2017-04-11 12:22:08 2017-04-13 13:25:17   
44       shipped      2018-06-04 16:44:48 2018-06-05 04:31:18   
103     invoiced      2018-08-03 17:44:42 2018-08-07 06:15:14   
128   processing      2017-09-03 14:22:03 2017-09-03 14:30:09   
154      shipped      2018-01-10 11:33:07 2018-01-11 02:32:30   

    order_delivered_carrier_date order_delivered_customer_date  \
6                            NaT                           NaT   
44           2018-06-05 14:32:00                           NaT   
103                          NaT                           NaT   
128                          NaT                           NaT   
154          2018-01-11 19:39:23                           NaT   

    order_estimated_delivery_date  
6                      2017-05-09  
44                     2018-06-28  
103                    2018-08-21  
128                    2017-10-03  
154                    2018-02-07

Какие есть статусы?

In [50]:
orders[orders.order_delivered_customer_date.isna()].order_status.unique()

array(['invoiced', 'shipped', 'processing', 'unavailable', 'canceled',
       'delivered', 'created', 'approved'], dtype=object)

У нас присутсвуют все виды статусов заказа, предположим, что чсть недоставлены по логичным причинам: не подтверждён, не выствлен счёт, не собран, не отгружен со склада или банально ещё не довезли, т.е. просто ещё не завершён. Также логично, что некоторые заказы на определённом этапе были отменены или стали недоступы. Т.о. один статус заказа "delivered" - не имеющий даты доставки, мы тое исключим, возможно, какой-то сбой. Такие заказы мы не будем учитывать.

In [51]:
orders[orders.order_delivered_customer_date.isna()].query("order_status != 'delivered'").shape[0]

2957

2957 заказов недоставлены по разным причинам. Мы взяли все заказы без даты доставки и исключили из полученных заказы со статусом "доставлен".

Но что, если мы вернёмся к первому варианту и пойдём обратным путём: сперва возьмём все заказы со статусом, отличным от "доставлен" и оттуда удалм все заказы, где оказалась дата доставки.

In [52]:
orders.query("order_status != 'delivered'")[orders.query("order_status != 'delivered'").order_delivered_customer_date.isna()].shape[0]

2957

Всё сошлось, т.к. 6 заказов без даты доставки имеют статус "отменён"

In [53]:
#создадим новый ДФ с недоставлнными заказами
no_delivered = orders[orders.order_delivered_customer_date.isna()].query("order_status != 'delivered'")

In [54]:
#приведём даты к месяцам, т.к. нас интересуют данным в разрезе месяца
no_delivered.order_purchase_timestamp = no_delivered.order_purchase_timestamp.dt.strftime('%Y-%m')

In [55]:
#посчитаем кол-во недоставленных заказов по каждой причине за каждый месяц, далее среднее значение для каждой причины
no_delivered \
    .groupby(["order_purchase_timestamp", "order_status"], as_index=False) \
    .agg({"customer_id": "count"}) \
    .groupby("order_status") \
    .agg({"customer_id": "mean"}) \
    .round() \
    .sort_values("customer_id")

customer_id
order_status             
approved              1.0
created               2.0
invoiced             15.0
processing           16.0
canceled             26.0
unavailable          29.0
shipped              48.0

# 3. По каждому товару определить, в какой день недели товар чаще всего покупается. (5 баллов)

In [56]:
#сделаем ДФ с покупками и датами
proba = orders.merge(order_items)

In [57]:
#псмотрим, что выло
proba.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  order_item_id  \
0                    2017-10-18              1   
1                    2018-08-13              1   
2                    2018-09-04              1   
3                    2017-12-15              1   
4                    2018-02-26              1   

                         product_id                         seller_id  \
0  87285b34884572647811a353c7ac498a  3504c0cb71d7fa48d967e0e4c94d59d9   
1  595fac2a385ac33a80bd5114aec74eb8  289cdb325fb7e7f891c38608bf9e0962   
2  aa4383b373c6aca5d8797843e5594415  4869f7a5dfa277a7dca6462dcf3b52b2   
3  d0b61bfb1de832b15ba9d266ca96e5b0  66922902710d126a0e7d26b0e3805106   
4  65266b2da20d04dbe00c5c2d3bb7859e  2c9e548be18521d1c43cde1c582c6de8   

   shipping_limit_date   price  freight_value  
0  2017-10-06 11:07:15   29.99           8.72  
1  2018-07-30 03:24:27  118.70          22.76  
2  2018-08-13 08:55:23  159.90          19.22  
3  2017-11-23 19:45:59   45.00          27.20  
4  2018-02-19 20:31:37   19.90           8.72

In [58]:
#Рашее мы уже определили какие статусы мы считаем покупкой, оставим только их
proba_bye = proba.query("order_status == ['processing', 'shipped', 'delivered']")

Единственная колонка с датами  - дата создания заказа, определим по ней день недели

In [59]:
proba_bye.order_purchase_timestamp = proba_bye.order_purchase_timestamp.dt.day_name()

/opt/tljh/user/lib/python3.7/site-packages/pandas/core/generic.py:5165: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [60]:
proba_bye \
    .groupby(["product_id", "order_purchase_timestamp"], as_index=False) \
    .agg({"customer_id": "count"}) \
    .sort_values(by="customer_id", ascending=False) \
    .drop_duplicates(subset='product_id')
#не забудь удалить колонку customer_id

product_id order_purchase_timestamp  customer_id
15766  422879e10f46682990de24d770e7f83d                Wednesday           93
36178  99a4788cb24856965c36a24e339b6058                   Monday           92
40679  aca2eb7d00ea1a7b8ebd4e68314663af                 Thursday           89
19944  53b36df67ebb7c41585e8d54d6772e08                  Tuesday           76
13422  389d119b48cf3043d311335e499d9c6b                 Thursday           67
...                                 ...                      ...          ...
23811  6435be64ad92f52631170f9a92718a99                   Sunday            1
23813  64367bed3ed96fd9381676fbe33b6443                   Sunday            1
23814  643a4049ae3e6df63858575f28cd7b21                   Friday            1
23816  643a66b1dc5dad3de6cb5a41549e72f1                   Sunday            1
60676  fffe9eeff12fcbd74a2f2b007dde0c58                Wednesday            1

[32614 rows x 3 columns]

# 4. Сколько у каждого из пользователей в среднем покупок в неделю (по месяцам)? Не стоит забывать, что внутри месяца может быть не целое количество недель. Например, в ноябре 2021 года 4,28 недели. И внутри метрики это нужно учесть. (8 баллов)  

In [61]:
customers_orders_items = customers_orders.merge(order_items)
#обединили все три ДФ в один

In [62]:
#приведём даты к месяцам, т.к. нас интересуют данные в разрезе месяца
customers_orders_items.order_purchase_timestamp = customers_orders_items.order_purchase_timestamp.dt.strftime('%Y-%m')

In [63]:
customers_orders_items_proba = customers_orders_items \
    .groupby(["customer_unique_id", "order_purchase_timestamp"], as_index=False) \
    .agg({"customer_state":"count"})
#выводим по каждому пользователю кол-во покупок в каждый месяц

In [64]:
customers_orders_items_proba["Mon"] = customers_orders_items_proba.order_purchase_timestamp
#делаем ещё 1 столбец с датой

In [65]:
customers_orders_items_proba["Mon"] = pd.to_datetime(customers_orders_items_proba["Mon"]).dt.month
#превращаем его в номер месяца

In [66]:
def week(mon):
        if mon == 2:
            x = 28
        elif mon == 4:
            x = 30
        elif mon == 6:
            x = 30
        elif mon == 9:
            x = 30
        elif mon == 11:
            x = 30
        else:
            x = 31
        return x 
    
#Сделали функцию, которая скажет солько дней в каждо из месяцев
#Почему не работает or??????

In [67]:
customers_orders_items_proba.Mon = customers_orders_items_proba.Mon.apply(week)
#применили

In [68]:
customers_orders_items_proba.Mon = (customers_orders_items_proba.Mon / 7)
#мне кажетс или я иду по сложному пути? в datetime есть функция, котора может сразу сказать, сколько в месяце дней?
#ну, я по крайней мере умею пользоваться циклами и создавать функции. Ну, если не считать того, что or работает не так, как хотелось бы

In [69]:
customers_orders_items_proba["tyt_nazvanie_metriki"] = (customers_orders_items_proba.customer_state / customers_orders_items_proba.Mon)

In [70]:
customers_orders_items_proba.head()

customer_unique_id order_purchase_timestamp  customer_state  \
0  0000366f3b9a7992bf8c76cfdf3221e2                  2018-05               1   
1  0000b849f77a49e4a4ce2b2a4ca5be3f                  2018-05               1   
2  0000f46a3911fa3c0805444483337064                  2017-03               1   
3  0000f6ccb0745a6a4b88665a16c9f078                  2017-10               1   
4  0004aac84e0df4da2b147fca70cf8255                  2017-11               1   

        Mon  tyt_nazvanie_metriki  
0  4.428571              0.225806  
1  4.428571              0.225806  
2  4.428571              0.225806  
3  4.428571              0.225806  
4  4.285714              0.233333

# 5. Используя pandas, проведи когортный анализ пользователей. В период с января по декабрь выяви когорту с самым высоким retention на 3й месяц. Описание подхода можно найти тут. (10 баллов)

Вспомним как рассчитывается Retention: Retention дня X — это N% процентов пользователей, которые вернутся к продукту в день X. Иными словами, если в какой-то конкретный день (день 0) пришло 100 новых пользователей, а на 1-ый день вернулось 15, то Retention 1-го дня составит 15 / 100 = 15%.

In [71]:
customers_orders_5 = customers_orders
#создадим датасет под задачу

Итого, будем считать, что все заказы со статусами: processing, shipped, delivered - оплачены и не отменены, соответственно, пользователь совершил покупку.

In [72]:
customers_orders_5 = customers_orders_5.dropna(axis='index', how='any', subset=['order_approved_at'])
#Удаляем пропуски

In [73]:
customers_orders_5["cohort"] = customers_orders_5.groupby('customer_unique_id')['order_approved_at'].transform('min')
#Создадим колонку с датой первой покупки

/opt/tljh/user/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [74]:
#приведём даты к месяцам, т.к. нас интересуют данные в разрезе месяца
customers_orders_5.order_purchase_timestamp = customers_orders_5.order_purchase_timestamp.dt.strftime('%Y-%m')
customers_orders_5.order_approved_at = customers_orders_5.order_approved_at.dt.strftime('%Y-%m')
customers_orders_5.first_buy = customers_orders_5['cohort'].dt.strftime('%Y-%m')

/opt/tljh/user/lib/python3.7/site-packages/pandas/core/generic.py:5165: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value
/opt/tljh/user/lib/python3.7/site-packages/ipykernel_launcher.py:4: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  after removing the cwd from sys.path.


In [75]:
customers_orders_5 = customers_orders_5.query("order_status == ['processing', 'shipped', 'delivered']") \
    .query("'2017-01' <= cohort <= '2017-12'") \
    .groupby(["cohort", "order_approved_at"], as_index=False) \
    .agg({"customer_unique_id": "nunique"})
#фильтруем по условиям, группируем даты и считаем кол-во уникальных айди

In [76]:
customers_orders_5["cohort"] = pd.to_datetime(customers_orders_5["cohort"]).dt.to_period("m")
customers_orders_5["order_approved_at"] = pd.to_datetime(customers_orders_5["order_approved_at"]).dt.to_period("m")
#превращаем в номер месяца

In [77]:
customers_orders_5['period'] = (customers_orders_5.order_approved_at - customers_orders_5.first_buy).apply(attrgetter("n"))
#считаем месяц покупки, начинаем с нулевого - месяц первой покупки

AttributeError: 'DataFrame' object has no attribute 'first_buy'

In [62]:
customers_orders_5_p = customers_orders_5.pivot(index='cohort', columns = 'period', values = 'customer_unique_id')
#считаем

In [63]:
customers_orders_5_p

cohort         0     1     2     3     4     5     6     7     8    9     10  \
first_buy                                                                      
2017-01     705.0   4.0   2.0   1.0   3.0   1.0   3.0   1.0   1.0  NaN   2.0   
2017-02    1661.0   4.0   4.0   3.0   7.0   2.0   4.0   3.0   2.0  4.0   2.0   
2017-03    2577.0  13.0   9.0   9.0   9.0   4.0   4.0   8.0   8.0  3.0   9.0   
2017-04    2286.0  14.0   5.0   4.0   7.0   6.0   8.0   6.0   8.0  4.0   5.0   
2017-05    3527.0  19.0  18.0  14.0  11.0  12.0  15.0   6.0   9.0  9.0  10.0   
2017-06    3102.0  15.0  11.0  12.0   8.0  10.0  12.0   7.0   4.0  7.0  10.0   
2017-07    3768.0  20.0  13.0   8.0  13.0   7.0  12.0   4.0   7.0  9.0   9.0   
2017-08    4135.0  28.0  15.0  11.0  16.0  22.0  11.0  11.0   5.0  7.0  10.0   
2017-09    4081.0  27.0  20.0  14.0  17.0  11.0   9.0   9.0  13.0  7.0  11.0   
2017-10    4343.0  30.0  13.0   4.0  10.0   8.0  11.0  16.0  12.0  9.0   9.0   
2017-11    7020.0  39.0  27.0  11.0  14.0  11.0   9.0  14.0   8.0  5.0   NaN   
2017-12    5575.0  15.0  15.0  18.0  14.0  11.0   9.0   2.0  12.0  NaN   NaN   

cohort       11   12    13   14   15   16   17   19  
first_buy                                            
2017-01     2.0  5.0   3.0  1.0  1.0  2.0  3.0  1.0  
2017-02     5.0  3.0   3.0  2.0  1.0  1.0  4.0  NaN  
2017-03     3.0  6.0   3.0  4.0  6.0  2.0  4.0  NaN  
2017-04     2.0  2.0   1.0  2.0  2.0  5.0  NaN  NaN  
2017-05    12.0  7.0   2.0  7.0  9.0  NaN  NaN  NaN  
2017-06    11.0  5.0   4.0  6.0  NaN  NaN  NaN  NaN  
2017-07    11.0  5.0  10.0  NaN  NaN  NaN  NaN  NaN  
2017-08     7.0  5.0   NaN  NaN  NaN  NaN  NaN  NaN  
2017-09     3.0  NaN   NaN  NaN  NaN  NaN  NaN  NaN  
2017-10     NaN  NaN   NaN  NaN  NaN  NaN  NaN  NaN  
2017-11     NaN  NaN   NaN  NaN  NaN  NaN  NaN  NaN  
2017-12     NaN  NaN   NaN  NaN  NaN  NaN  NaN  NaN

In [64]:
round(customers_orders_5_p.div(customers_orders_5_p.iloc[:,0], axis = 0).mul(100), 2)[2].idxmax()
#выводим когорту с максимаьным значениеем на третий месяц (индекс 2!)

Period('2017-05', 'M')

# 6. Часто для качественного анализа аудитории использую подходы, основанные на сегментации. Используя python, построй RFM-сегментацию пользователей, чтобы качественно оценить свою аудиторию. В кластеризации можешь выбрать следующие метрики: R - время от последней покупки пользователя до текущей даты, F - суммарное количество покупок у пользователя за всё время, M - сумма покупок за всё время. Подробно опиши, как ты создавал кластеры. Для каждого RFM-сегмента построй границы метрик recency, frequency и monetary для интерпретации этих кластеров. Пример такого описания: RFM-сегмент 132 (recency=1, frequency=3, monetary=2) имеет границы метрик recency от 130 до 500 дней, frequency от 2 до 5 заказов в неделю, monetary от 1780 до 3560 рублей в неделю. Описание подхода можно найти тут. (35 баллов)

In [119]:
customers_orders_6 = customers_orders_items
#создадим датасет для 6 задачи

In [120]:
customers_orders_6 = customers_orders_6.dropna()
#Удалим все пропуски

order_approved_at - дата оплаты, значит продажа состоялсь, будем отталкиватся от уже имеющейся логики

In [121]:
#customers_orders_6.order_approved_at = pd.to_datetime(customers_orders_6.order_approved_at)

In [122]:
last_day = customers_orders_6.order_approved_at.max()

Допустим, мы выгрузили отчёт вчера, т.о. в строчке выше мы нашли вчерашнюю дату, добавим день, чтобы у нас получилось условное СЕГОДНЯ.

In [123]:
today = last_day + timedelta(days=1)

Создаём таблицу RFM

R - время от последней покупки пользователя до текущей даты
F - суммарное количество покупок у пользователя за всё время
M - стоимость покупок за всё время

In [124]:
RFM = customers_orders_6 \
    .groupby('customer_unique_id') \
    .agg({'order_approved_at': lambda x: (today - x.max()).days, 'product_id': lambda x: len(x), 'price': lambda x: x.sum()}) \
    .rename(columns={'order_approved_at': 'R', 'product_id': 'F', 'price': 'M'})

Мы будем разбивать на диапазон от 1 до 5: до 20, 20-40, 40-60, 60-80 и выше 80 %

In [125]:
RFM_seg = RFM.quantile(q=[0.20, 0.40, 0.60, 0.80]).to_dict()

In [ ]:
Напишем функцию, которая сегментирует покупателей

In [126]:
def to_seg (a,b,c):
    if a <= c[b][0.20]:
        return 1
    elif a <= c[b][0.40]:
        return 2
    elif a <= c[b][0.60]:
        return 3
    elif a <= c[b][0.80]:
        return 4
    else:
        return 5

Применяем функцию для каждого кластера

In [127]:
RFM['R_seg'] = RFM['R'].apply(to_seg, args=('R', RFM_seg, ))
RFM['F_seg'] = RFM['F'].apply(to_seg, args=('F', RFM_seg, ))
RFM['M_seg'] = RFM['M'].apply(to_seg, args=('M', RFM_seg, ))

Склеиваем итог в одно поле

In [128]:
RFM['RFMseg'] = RFM.R_seg.map(str) + RFM.F_seg.map(str) + RFM.M_seg.map(str)

In [129]:
RFM

R  F        M  R_seg  F_seg  M_seg RFMseg
customer_unique_id                                                           
0000366f3b9a7992bf8c76cfdf3221e2  112  1   129.90      2      1      4    214
0000b849f77a49e4a4ce2b2a4ca5be3f  114  1    18.90      2      1      1    211
0000f46a3911fa3c0805444483337064  537  1    69.00      5      1      2    512
0000f6ccb0745a6a4b88665a16c9f078  321  1    25.99      4      1      1    411
0004aac84e0df4da2b147fca70cf8255  288  1   180.00      4      1      5    415
...                               ... ..      ...    ...    ...    ...    ...
fffcf5a5ff07b0908bd4e2dbc735a684  447  2  1570.00      5      5      5    555
fffea47cd6d3cc0a88bd621562a9d061  262  1    64.89      3      1      2    312
ffff371b4d645b6ecea244b27531430a  568  1    89.90      5      1      3    513
ffff5962728ec6157033ef9805bacc48  119  1   115.00      2      1      4    214
ffffd2657e2aad2907e67c3e9daecbeb  483  1    56.99      5      1      2    512

[93342 rows x 7 columns]

Клиенты по RFM:
555 — Лояльные и активные: покупали недавно, покупают часто и много платят;
111 — Отток: покупали давно, покупают редко и мало платят;
R=[1, 2], F=[4, 5], M=[4, 5] — Ляльные, потерявшие активность: покупали давно, но часто и много платили